### PR TITLE
Add CBMC proof for DHCPProcess IPv4

### PR DESCRIPTION
**Pull request by tony-josi-aws**
Description
-----------
This PR adds CBMC proof to `DHCPProcess()`

Test Steps
-----------
```
cd test/cbmc/proof/
python3 prepare.py
cd <test_case>
make
```

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json [hunk 1]
```diff
@@ -36,10 +36,10 @@
   # The number of times GetNetworkBufferWithDescriptor can be allowed to fail
```
Comment (moninom1) [inline test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json:36]: Comments are not supported in json file

File: test/cbmc/proofs/DHCP/DHCPProcessEndPoint/Makefile.json [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# http://aws.amazon.com/freertos
+# http://www.FreeRTOS.org
+#
```
Comment (moninom1) [inline test/cbmc/proofs/DHCP/DHCPProcessEndPoint/Makefile.json:27]: We have decided to not add comments in json file, please check the file format once in the doc.

File: test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json (hunk 1 above)
Comment (tony-josi-aws) [reply test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json:36]: We can remove the copyright, but in my opinion its better to keep these comments as its helpful for future maintenance and the python script is taking care of the comments for now.

Comment (moninom1) [reply test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json:36]: Yeah we can skip for existing comments

File: test/cbmc/proofs/DHCP/DHCPProcessEndPoint/DHCPProcessEndPoint_harness.c [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    return xRet;
+}
+
+
+/*We assume that the pxGetNetworkBufferWithDescriptor function is implemented correctly and returns a valid data structure. */
+/*This is the mock to mimic the correct expected behavior. If this allocation fails, this might invalidate the proof. */
+NetworkBufferDescriptor_t * pxGetNetworkBufferWithDescriptor( size_t xRequestedSizeBytes,
+                                                              TickType_t xBlockTimeTicks )
+{
+    NetworkBufferDescriptor_t * pxNetworkBuffer = ( NetworkBufferDescriptor_t * ) malloc( sizeof( NetworkBufferDescriptor_t ) );
```
Comment (moninom1) [inline test/cbmc/proofs/DHCP/DHCPProcessEndPoint/DHCPProcessEndPoint_harness.c:114]: Please use safeMalloc

File: test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json [hunk 4]
```diff
@@ -1,31 +1,3 @@
-#
```
Comment (moninom1) [inline test/cbmc/proofs/DHCP/DHCPProcess/Makefile.json:1]: Remove copyright

**Comment by moninom1** [comment]
Can you mention which function along with DHCPProcess is getting validated in this proof?

**Comment by tony-josi-aws** [comment]
> Can you mention which function along with DHCPProcess is getting validated in this proof?

@moninom1, only DHCPProcess is validated. The changes to vDHCPProcessEndPoint is just refactoring changes.

File: test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c [hunk 5]
```diff
@@ -111,25 +115,55 @@ BaseType_t vSocketBind( FreeRTOS_Socket_t * pxSocket,
...
                                                               TickType_t xBlockTimeTicks )
 {
-    NetworkBufferDescriptor_t * pxNetworkBuffer = ( NetworkBufferDescriptor_t * ) malloc( sizeof( NetworkBufferDescriptor_t ) );
+    NetworkBufferDescriptor_t * pxNetworkBuffer = ( NetworkBufferDescriptor_t * ) safeMalloc( sizeof( NetworkBufferDescriptor_t ) );
 
     __CPROVER_assume( pxNetworkBuffer != NULL );
     __CPROVER_assume( xRequestedSizeBytes > ( dhcpFIRST_OPTION_BYTE_OFFSET + sizeof( MACAddress_t ) + ipIP_TYPE_OFFSET ) );
 
-    pxNetworkBuffer->pucEthernetBuffer = ( ( uint8_t * ) malloc( xRequestedSizeBytes + ( ipIP_TYPE_OFFSET ) ) ) + ipIP_TYPE_OFFSET;
+    pxNetworkBuffer->pucEthernetBuffer = ( ( uint8_t * ) safeMalloc( xRequestedSizeBytes + ( ipIP_TYPE_OFFSET ) ) );
```
Comment (moninom1) [inline test/cbmc/proofs/DHCP/DHCPProcess/DHCPProcess_harness.c:123]: This needs to be corrected

**Review by moninom1 — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Pull request merged**